File: generated-code/php/codecraft/TcpReadWrite/Codegame/ClientMessage.php
<?php

namespace Codegame {
    require_once 'Codegame/DebugCommand.php';
    require_once 'Model/Action.php';
    require_once 'Stream.php';

    /**
     * Message sent from client
     */
    abstract class ClientMessage
    {
        /**
         * Write ClientMessage to output stream
         */
        abstract function writeTo(\OutputStream $stream): void;

        /**
         * Read ClientMessage from input stream
         */
        static function readFrom(\InputStream $stream): ClientMessage
        {
            $tag = $stream->readInt32();
            if ($tag == \Codegame\ClientMessage\DebugMessage::TAG) {
                return \Codegame\ClientMessage\DebugMessage::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\ActionMessage::TAG) {
                return \Codegame\ClientMessage\ActionMessage::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\DebugUpdateDone::TAG) {
                return \Codegame\ClientMessage\DebugUpdateDone::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\RequestDebugState::TAG) {
                return \Codegame\ClientMessage\RequestDebugState::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Codegame\ClientMessage {
    /**
     * Ask app to perform new debug command
     */
    class DebugMessage extends \Codegame\ClientMessage
    {
        const TAG = 0;
    
        /**
         * Command to perform
         */
        public \Codegame\DebugCommand $command;
    
        function __construct(\Codegame\DebugCommand $command)
        {
            $this->command = $command;
        }
    
        /**
         * Read DebugMessage from input stream
         */
        public static function readFrom(\InputStream $stream): DebugMessage
        {
            $command = \Codegame\DebugCommand::readFrom($stream);
            return new DebugMessage($command);
        }
        
        /**
         * Write DebugMessage to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(DebugMessage::TAG);
            $this->command->writeTo($stream);
        }
    }

    /**
     * Reply for ServerMessage::GetAction
     */
    class ActionMessage extends \Codegame\ClientMessage
    {
        const TAG = 1;
    
        /**
         * Player's action
         */
        public \Model\Action $action;
    
        function __construct(\Model\Action $action)
        {
            $this->action = $action;
        }
    
        /**
         * Read ActionMessage from input stream
         */
        public static function readFrom(\InputStream $stream): ActionMessage
        {
            $action = \Model\Action::readFrom($stream);
            return new ActionMessage($action);
        }
        
        /**
         * Write ActionMessage to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(ActionMessage::TAG);
            $this->action->writeTo($stream);
        }
    }

    /**
     * Signifies finish of the debug update
     */
    class DebugUpdateDone extends \Codegame\ClientMessage
    {
        const TAG = 2;
    
    
        function __construct()
        {
        }
    
        /**
         * Read DebugUpdateDone from input stream
         */
        public static function readFrom(\InputStream $stream): DebugUpdateDone
        {
            return new DebugUpdateDone();
        }
        
        /**
         * Write DebugUpdateDone to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(DebugUpdateDone::TAG);
        }
    }

    /**
     * Request debug state from the app
     */
    class RequestDebugState extends \Codegame\ClientMessage
    {
        const TAG = 3;
    
    
        function __construct()
        {
        }
    
        /**
         * Read RequestDebugState from input stream
         */
        public static function readFrom(\InputStream $stream): RequestDebugState
        {
            return new RequestDebugState();
        }
        
        /**
         * Write RequestDebugState to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(RequestDebugState::TAG);
        }
    }
}

File: generated-code/haskell/codecraft/TcpReadWrite/src/Codegame/ClientMessage.hs
module Codegame.ClientMessage where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Codegame.DebugCommand (DebugCommand)
import Data.Int
import Model.Action (Action)

-- | Ask app to perform new debug command
data ClientMessageDebugMessage = ClientMessageDebugMessage {
    -- | Command to perform
    command :: DebugCommand }
    deriving Show

instance Trans ClientMessageDebugMessage where
    read = do
        command <- Trans.read
        return ClientMessageDebugMessage {
            command }

    write ClientMessageDebugMessage {
        command } = do
            Trans.write command

-- | Reply for ServerMessage::GetAction
data ClientMessageActionMessage = ClientMessageActionMessage {
    -- | Player's action
    action :: Action }
    deriving Show

instance Trans ClientMessageActionMessage where
    read = do
        action <- Trans.read
        return ClientMessageActionMessage {
            action }

    write ClientMessageActionMessage {
        action } = do
            Trans.write action

-- | Signifies finish of the debug update
data ClientMessageDebugUpdateDone = ClientMessageDebugUpdateDone { }
    deriving Show

instance Trans ClientMessageDebugUpdateDone where
    read = do
        return ClientMessageDebugUpdateDone { }

    write ClientMessageDebugUpdateDone { } = do
            return ()

-- | Request debug state from the app
data ClientMessageRequestDebugState = ClientMessageRequestDebugState { }
    deriving Show

instance Trans ClientMessageRequestDebugState where
    read = do
        return ClientMessageRequestDebugState { }

    write ClientMessageRequestDebugState { } = do
            return ()

-- | Message sent from client
data ClientMessage
    -- | Ask app to perform new debug command
    = DebugMessage ClientMessageDebugMessage
    -- | Reply for ServerMessage::GetAction
    | ActionMessage ClientMessageActionMessage
    -- | Signifies finish of the debug update
    | DebugUpdateDone ClientMessageDebugUpdateDone
    -- | Request debug state from the app
    | RequestDebugState ClientMessageRequestDebugState
    deriving Show

instance Trans ClientMessage where
    read = do
        tag :: Int32 <- Trans.read
        case tag of
            0 -> DebugMessage <$> Trans.read
            1 -> ActionMessage <$> Trans.read
            2 -> DebugUpdateDone <$> Trans.read
            3 -> RequestDebugState <$> Trans.read

    write (DebugMessage value) = do
        Trans.write (0 :: Int32)
        Trans.write value
    write (ActionMessage value) = do
        Trans.write (1 :: Int32)
        Trans.write value
    write (DebugUpdateDone value) = do
        Trans.write (2 :: Int32)
        Trans.write value
    write (RequestDebugState value) = do
        Trans.write (3 :: Int32)
        Trans.write value